File: Python/Queues/circularqueue.py
class CircularQueue:
    def __init__(self,capacity):
        self.storage = [None] * capacity
        self.head = 0
        self.tail = 0
        self.size = 0
        self.capacity = capacity

    def isFull(self):
        return self.capacity == self.size

    def enqueue(self,data):
        if(self.isFull()):
            raise Exception("Queue is Full")
        self.storage[self.tail] = data
        self.tail = (self.tail + 1) % self.capacity
        self.size += 1
    
    def dequeue(self):
        if(self.size == 0):
            return None
        data = self.storage[self.head]
        self.head = (self.head + 1) % self.capacity
        self.size -= 1
        return data

File: JavaScript/Queues/circularqueue.js
class CircularQueue{
    constructor(capacity){
        this.storage = [];
        this.head = 0;
        this.tail = 0;
        this.size = 0;
        this.capacity = capacity;
    }

    isFull(){
        return this.capacity === this.size;
    }

    enqueue(data){
        if(this.isFull())
            throw new Error("Queue is full");
        this.storage[this.tail] = data;
        this.tail = (this.tail + 1) % this.capacity;
        this.size++;
    }

    dequeue(){
        if(this.size === 0)
            return null;
        let data = this.storage[this.head];
        this.head = (this.head + 1) % this.capacity;
        this.size--;
        return data;
    }
}